discovery_child_development/notebooks/02_explore_concepts.py

```diff
@@ -11,164 +13,302 @@
 
-# +
-# Group the data by 'cluster_minilm' and aggregate the text for each cluster
-grouped_text = (
-    unique_concepts_df.groupby("cluster")
-    .agg(
-        {
-            "concept": ", ".join,
-            #'cluster': 'size'
-        }
-    )
+hdbscan_low_prob_docs_grouped = (
+    hdbscan_low_prob_docs.groupby("cluster")
+    .agg({"display_name": ";".join})
     .reset_index()
-)  # .rename(columns={'cluster': 'count'})
-# openalex_clusters.groupby('cluster_minilm')['text'].apply(' '.join).reset_index()
+)
+
+hdbscan_low_prob_docs_grouped
 
-grouped_text.head()
+# %%
+hdbscan_docs_grouped = hdbscan_docs_grouped.rename(
+    columns={"display_name": "high_prob_concepts"}
+)
+hdbscan_low_prob_docs_grouped = hdbscan_low_prob_docs_grouped.rename(
+    columns={"display_name": "low_prob_concepts"}
+)
 
-# +
-from sklearn.feature_extraction.text import TfidfVectorizer
+# %%
+hdbscan_docs_grouped = hdbscan_docs_grouped.merge(
+    hdbscan_low_prob_docs_grouped, on="cluster", how="left"
+)
+hdbscan_docs_grouped.head()
 
-# Initialize the TF-IDF vectorizer
-vectorizer = TfidfVectorizer(max_df=0.5, stop_words="english")
+# %%
+hdbscan_docs_grouped.to_csv("hdbscan_highest_prob_concept_groups.csv")
 
-# Fit and transform the aggregated text for each cluster
-tfidf_matrix = vectorizer.fit_transform(grouped_text["concept"])
+# %% [markdown]
+# # Most representative publication per concept
+#
+# It might be useful to know for each concept, which work got the highest score for that cluster.
 
-# Extract the top terms for each cluster
-top_terms = {}
-feature_names = vectorizer.get_feature_names_out()
-for i, row in enumerate(tfidf_matrix.toarray()):
-    top_terms[grouped_text.iloc[i]["cluster"]] = [
-        feature_names[index] for index in row.argsort()[-10:][::-1]
-    ]
+# %%
+most_relevant_work_per_concept = concepts_cleaned_df.loc[
+    concepts_cleaned_df.groupby("display_name")["score"].idxmax()
+][["concept_id", "display_name", "level", "openalex_id", "title", "score"]]
 
-top_terms_df = pd.DataFrame(top_terms).T.reset_index()
-top_terms_df.columns = ["cluster"] + [f"term_{i+1}" for i in range(10)]
-# -
+most_relevant_work_per_concept.head()
 
-top_terms_df
+# %%
+most_relevant_work_per_concept[
+    most_relevant_work_per_concept["display_name"] == "Emerging technologies"
+]
 
-top_terms_df.to_csv("top_terms_per_concept_cluster.csv", index=False)
+# %%
+most_relevant_work_per_concept[
+    most_relevant_work_per_concept["display_name"] == "Assistive technology"
+]
 
-# Observations:
-# * cluster 10 seems to be about assessment -> relevant to this ISS3 project
+# %%
+unique_concepts_df["core_peripheral"] = np.where(
+    unique_concepts_df["probability"] > 0.5, "core", "peripheral"
+)
 
-unique_concepts_df = unique_concepts_df.merge(
-    concepts_df, left_on="concept", right_on="display_name", how="left"
+unique_concepts_df = unique_concepts_df.rename(
+    columns={"probability": "cluster_probability"}
```

Comment:
Generally, I would suggest chaining all the dataframe operations, so that they are in one place, eg

```python
your_df = (
  your_df
  # add a comment
  .first_operation(params)
  # add another comment
  .second_operation(params)
  # and yet another comment
  .assign(new_column = lambda df: np.where(df.old_column < 0.5...)
  # and so on ...
)
```

However, I do appreciate that when developing within a notebook you might want to go step by step, and check intermediate results. However, afterwards, can quickly refactor into a compact chain.

Reply:
I think it's OK for now - can take note of this for the the next time!